ekf/log.c

```diff
@@ -9,152 +9,278 @@
 			}
 
 			pthread_mutex_lock(&ekflog_common.lock);
-			ekflog_common.tail += strlen(&ekflog_common.buff[ekflog_common.tail]);
-			pthread_cond_signal(&ekflog_common.queueEvent);
+
+			outBuff->dirty = false;
+			outBuff->size = 0;
+
+			outBuff = ekflog_nextBufferGet(outBuff);
+			pthread_cond_signal(&ekflog_common.buffEvent);
 		}
 	} while (ekflog_common.run != 0);
 
+	printf("Logging finished\n");
+	printf("Number of logs requests: %d\n", ekflog_common.logCnt);
+	printf("Lost logs: %d\n", ekflog_common.lost);
+
 	pthread_mutex_unlock(&ekflog_common.lock);
 
 	return NULL;
 }
 
 
-/*
- * Sets head to the beginning of new message slot.
- * Assumes that the length of a message is equal to MAX_MSG_LEN.
- * Tries to find place for the longest possible message.
- */
-static int ekflog_slotGet(void)
+static bool ekflog_actBuffWritable(void)
 {
-	if (ekflog_common.head < ekflog_common.tail) {
-		if (ekflog_common.tail - ekflog_common.head - 1 > MAX_MSG_LEN) {
-			/* Adding new message after previous one */
-			ekflog_common.head++;
-			return 0;
-		}
+	if (ekflog_common.actBuff->dirty == false) {
+		return true;
+	}
 
+	if ((ekflog_common.logFlags & EKFLOG_STRICT_MODE) != 0) {
+		/* Waiting for a place to insert logs */
+		do {
+			pthread_cond_wait(&ekflog_common.buffEvent, &ekflog_common.lock);
+		} while (ekflog_common.actBuff->dirty == true);
+
+		return true;
+	}
+
+	return false;
+}
+
+
+static int ekflog_write(void *msg, size_t msg_len)
+{
+	size_t remainingBuffSize;
+
+	pthread_mutex_lock(&ekflog_common.lock);
+
+	remainingBuffSize = BUFFS_CAPACITY - ekflog_common.actBuff->size;
+
+	if (remainingBuffSize < msg_len) {
+		/* Changing actual buffer for the next one */
+		ekflog_common.actBuff->dirty = true;
+		ekflog_common.actBuff = ekflog_nextBufferGet(ekflog_common.actBuff);
+		ekflog_common.actBuff->head = ekflog_common.actBuff->buff;
+		pthread_cond_signal(&ekflog_common.buffEvent);
+	}
+
+	if (!ekflog_actBuffWritable()) {
+		/* Dropping the log */
+		ekflog_common.lost++;
+		pthread_mutex_unlock(&ekflog_common.lock);
 		return -1;
 	}
 
-	/* ekflog_common.head >= ekflog_common.tail */
+	memcpy(ekflog_common.actBuff->head, msg, msg_len);
+	ekflog_common.actBuff->head += msg_len;
+	ekflog_common.actBuff->size += msg_len;
+
+	pthread_mutex_unlock(&ekflog_common.lock);
+
+	return 0;
+}
+
+
+static int ekflog_writeLogPrefix(char *msg_buf, char msgType, time_t timestamp)
+{
+	uint64_t time = (uint64_t)timestamp;
+
+	ekflog_common.logCnt++;
+
+	memcpy(msg_buf, &ekflog_common.logCnt, sizeof(ekflog_common.logCnt));
+	msg_buf[sizeof(ekflog_common.logCnt)] = msgType;
+	memcpy(&msg_buf[sizeof(ekflog_common.logCnt) + sizeof(msgType)], &time, sizeof(time));
+
+	return sizeof(uint32_t) + sizeof(char) + sizeof(uint64_t);
+}
+
+
+static inline int ekflog_addLogField(char *msg_buf, const void *data, size_t dataSize)
+{
+	memcpy(&msg_buf, data, dataSize);
+	return dataSize;
+}
+
+
+extern int ekflog_timeWrite(time_t timestamp)
```

Comment:
`extern` redundant, also with other functions
```suggestion
int ekflog_timeWrite(time_t timestamp)
```

Reply:
We are using `extern` in all other header file, so for consistency I leave as it is

Reply:
Ok, I agree. That's my fault. Fixed in separate PR -> https://github.com/phoenix-pilot/phoenix-pilot-core/pull/251